doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,209 @@
+are safe and so are operators that can re-construct their in-memory state by reading a snapshot of
+`C` (not `RAW`). The upsert envelope is an instance of the latter case.
+
+### Adding arbitrary SQL to `TRANSFORM`
+
+Given the constraints of the transformation function that can be ran during ingestion we can explore
+a few different avenues to incorporate an arbitrary query `Q` as a transformation step. Every query
+in materialize, after planning and optimizing, can be decomposed into a `MapFilterProject` (MFP from
+now on) part (`Q_mfp`) and a relation expression part (`Q_expr`). Depending on the query any of
+these can be trivial (i.e the identity).
+
+This decomposition is relevant to what we’re trying to accomplish because `Q_mfp` has the property
+that it’s stateless (state for temporal filters discussed below) and can be evaluated without having
+access to input snapshots whereas `Q_expr` might include arrangements and therefore will need an
+input snapshot before it can resume.
+
+With this decomposition we can instantiate two dataflows:
+
+1. `C_intermediate = PERSIST(Q_mfp(RAW))`
+2. `C = PERSIST(Q_expr(C_intermediate))`
+
+Where the first is scheduled in the storage timely cluster and the second in the compute timely
+cluster. In a future where the two timely clusters get unified this could be a single big dataflow.
+
+### Handling MFPs with temporal filters
+
+Evaluation of `MFP`s is parameterized by the `Row` that we want to map, filter and project and the
+time `t` that this evaluation is happening at. When the `MFP` contains no temporal filters the
+output is a single transformed `Row` at the same time `t`. However, when there is a temporal filter
+involved there is a possibility that we get `Row`s that happen in the future. This is because for
+each `Row` the temporal filter will calculate a `lower_ts` and an `upper_ts` which define the
+validity period for the row. Then these differential updates will be produced: `(row, lower_ts, +1)`
+and `(row, upper_ts, -1)`. The `lower_ts` and `upper_ts` timestamps are entirely dependent on the
+expression the user typed and can be arbitrarily in the future.. This is best depicted in this
+diagram:
+
+![Temporal filter application](https://github.com/MaterializeInc/materialize/raw/9c79faf5d58b3a97b11f982fa4f6e7736c864922/doc/developer/design/static/temporal_filters.png)
+
+This is a problem because when the `RAW` colletion produces the differential update `(row, t, +1)`
+we might have to emit it at some future time `t_temporal >> t` but if we crash and restart we might
+have to resume from some time `t_resume > t && t_resume < t_temporal`. In this case we will entirely
+miss the original update from `RAW` and we will fail to produce the differential update that we
+would have produced had we not crashed.
+
+In order to solve this we will need to decompose `Q_mfp` into three parts: `Q_safemfp`,
```

Comment:
I have a counter proposal, and I am curious what you think about it. 

We instead break down `Q_mfp` into two parts: `Q_mfp_pre` and `Q_mfp_post`, where the latter contains any temporal lower bounds (and columns needed to support them), and the former contains everything else. We have already machinery to slice out parts of a MFP (used in join planning, based on column availability). It seems like the same thing would work here: break the MFP into two parts, probably using the same idea as here of "compute the lower bound as a column, but then mark it as unavailable to the predicate".

Reply:
I think in practice we will have to break it down into three parts anyway (I updated the document to reflect the new breakdown logic) because we need to have at least one MFP that can be evaluated both on its own output (for recovering the state from `Q_expiration(snapshot(C_intermediate))`) and on its input (for computing normal updates with `Q_expiration(Q_safemfp(RAW))`). This means that that MFP must contain only predicates, or at least have an identical input and output schema whereas `Q_safemfp` might have an arbitrary output schema.